File: java/name-cache/src/main/java/org/mydotey/namecache/NameCacheNode.java
package org.mydotey.namecache;

import java.util.concurrent.ConcurrentHashMap;

/**
 * @author koqizhao
 *
 * Jul 20, 2018
 */
public class NameCacheNode<K, V> {

    private K key;
    private V value;
    private ConcurrentHashMap<K, NameCacheNode<K, V>> children;

    public NameCacheNode() {
        this(null);
    }

    public NameCacheNode(K key) {
        this.key = key;
        children = new ConcurrentHashMap<K, NameCacheNode<K, V>>();
    }

    public K getKey() {
        return key;
    }

    public void setKey(K key) {
        this.key = key;
    }

    public V getValue() {
        return value;
    }

    public void setValue(V value) {
        this.value = value;
    }

    public ConcurrentHashMap<K, NameCacheNode<K, V>> getChildren() {
        return children;
    }

    public void setChildren(ConcurrentHashMap<K, NameCacheNode<K, V>> children) {
        this.children = children;
    }
}

File: dotnet/src/MyDotey.NameCache/NameCacheNode.cs
using System;
using System.Collections.Generic;
using System.Collections.Concurrent;
using System.Linq;
using System.Text;

namespace MyDotey.NameCache
{
    public class NameCacheNode<K, V>
    {
        public K Key { get; set; }
        public V Value { get; set; }
        public ConcurrentDictionary<K, NameCacheNode<K, V>> Children { get; set; }

        public NameCacheNode()
            : this(default(K))
        {
        }

        public NameCacheNode(K key)
        {
            Key = key;
            Children = new ConcurrentDictionary<K, NameCacheNode<K, V>>();
        }
    }
}
